src/log-store/src/kafka/producer/batch.rs

```diff
@@ -0,0 +1,222 @@
+    tag: A::Tag,
+}
+
+impl<A: aggregator::Aggregator> ResultHandle<A> {
+    /// Waits for [`BatchFlushState`] changed(exclude [`BatchFlushState::Init`]) and returns the  [`AggregatedStatus`].
+    ///
+    /// # Cancel safety
+    ///
+    /// This method is cancel safe. If you use it as the event in a
+    /// [`tokio::select!`] statement and some other branch
+    /// completes first, then it is guaranteed that no values have been marked
+    /// seen by this call to `changed`.
+    pub(crate) async fn wait(&mut self) -> Result<Arc<AggregatedStatus<A>>> {
+        self.receiver
+            .changed()
+            .await
+            .context(error::StateRecvSnafu)?;
+
+        let status = &*self.receiver.borrow();
+        match status {
+            BatchFlushState::Init => unreachable!(),
+            BatchFlushState::Result(status) => Ok(status.clone()),
+            BatchFlushState::Err(err) => Err(err.clone()).context(error::FlushErrSnafu),
+        }
+    }
+
+    /// Return the demuxed result of the produce() call.
+    pub(crate) fn result(
+        self,
+        status: Arc<AggregatedStatus<A>>,
+    ) -> Result<<A as aggregator::AggregatorStatus>::Status> {
+        status
+            .status_deagg
+            .deaggregate(&status.aggregated_status, self.tag)
+            .context(error::DeaggregateStatusSnafu)
+    }
+}
+
+/// A [`BatchBuilder`] uses an [`Aggregator`] to construct maximally large batch
+/// of writes, and returning a [`ResultHandle`] for callers to demux the result.
+pub(crate) struct BatchBuilder<A>
+where
+    A: aggregator::Aggregator,
+{
+    aggregator: A,
+    /// Sends the flush `results` to [ResultHandle]s.
+    sender: watch::Sender<BatchFlushState<A>>,
+    /// The receiver of `results`.
+    receiver: watch::Receiver<BatchFlushState<A>>,
+}
+
+pub(crate) struct FlushRequest<A: Aggregator> {
+    pub(crate) batch: Vec<Record>,
+    pub(crate) status_deagg: A::StatusDeaggregator,
+    /// Sends the flush `results` to [ResultHandle]s.
+    pub(crate) sender: watch::Sender<BatchFlushState<A>>,
+}
+
+impl<A: aggregator::Aggregator> BatchBuilder<A> {
+    pub(crate) fn new(aggregator: A) -> Self {
+        let (sender, receiver) = watch::channel(BatchFlushState::Init);
+
+        Self {
+            aggregator,
+            sender,
+            receiver,
+        }
+    }
+
+    pub(crate) fn try_push(
+        &mut self,
+        data: A::Input,
+    ) -> Result<TryPush<A::Input, ResultHandle<A>>> {
+        match self
+            .aggregator
+            .try_push(data)
+            .context(error::AggregateInputSnafu)?
+        {
+            TryPush::NoCapacity(data) => Ok(TryPush::NoCapacity(data)),
+            TryPush::Aggregated(tag) => Ok(TryPush::Aggregated(ResultHandle {
+                receiver: self.receiver.clone(),
+                tag,
+            })),
+        }
+    }
+
+    pub(crate) fn next_batch(self) -> (Self, Result<Option<FlushRequest<A>>>) {
```

Comment:
Could we simply use `&mut self` as input and return a `Result<Option<FlushRequest<A>>>`?